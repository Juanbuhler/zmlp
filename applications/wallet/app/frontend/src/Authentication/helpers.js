export const USER = 'user'

export const getUser = () => {
  try {
    return JSON.parse(localStorage.getItem(USER)) || {}
  } catch (error) {
    return {}
  }
}

export const clearUser = () => {
  localStorage.removeItem(USER)
}

export const storeUser = ({ user }) => {
  localStorage.setItem(USER, JSON.stringify(user))
}

export const authenticateUser = ({ setUser }) => async ({
  username,
  password,
}) => {
  const response = await fetch('/api/v1/login', {
    method: 'POST',
    headers: { 'content-type': 'application/json;charset=UTF-8' },
    body: JSON.stringify({ username, password }),
  })

  const user = await response.json()

  storeUser({ user })

  setUser(user)
}

export const logout = ({ setUser }) => async () => {
  await fetch('/api/v1/logout', {
    method: 'POST',
    headers: { 'content-type': 'application/json;charset=UTF-8' },
  })

  clearUser()

  setUser({})
}

export const fetcher = ({ setUser }) => async (...args) => {
  const response = await fetch(...args)

  if (response.status === 401) {
    clearUser()

    setUser({})

    return {}
  }

  return response.json()
}
