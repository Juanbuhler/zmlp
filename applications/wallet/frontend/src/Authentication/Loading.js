import { spacing } from '../Styles'

import Loading from '../Loading'

const AuthenticationLoader = () => {
  return (
    <div css={{ height: '100vh', display: 'flex' }}>
      <div css={{ flex: 1, display: 'flex', padding: spacing.spacious }}>
        <Loading />
      </div>
    </div>
  )
}

export default AuthenticationLoader
