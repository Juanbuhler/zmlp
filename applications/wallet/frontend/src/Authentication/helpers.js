import Router from 'next/router'

export const authenticateUser = ({ setUser, setErrorMessage }) => async ({
  username,
  password,
  idToken,
}) => {
  setErrorMessage('')

  const response = await fetch('/api/v1/login/', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json;charset=UTF-8',
      ...(idToken ? { Authorization: `Bearer ${idToken}` } : {}),
    },
    ...(idToken ? {} : { body: JSON.stringify({ username, password }) }),
  })

  if (response.status === 401) {
    return setErrorMessage('Invalid email or password.')
  }

  if (response.status !== 200) {
    return setErrorMessage('Network error.')
  }

  const user = await response.json()

  return setUser({ user: { ...user, projectId: '' } })
}

export const logout = ({ googleAuth, setUser }) => async () => {
  googleAuth.signOut()

  const { csrftoken } = Object.fromEntries(
    document.cookie.split(/; */).map(c => {
      const [key, ...v] = c.split('=')
      return [key, decodeURIComponent(v.join('='))]
    }),
  )

  await fetch('/api/v1/logout/', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json;charset=UTF-8',
      'X-CSRFToken': csrftoken,
    },
  })

  setUser({ user: null })

  Router.push('/')
}
