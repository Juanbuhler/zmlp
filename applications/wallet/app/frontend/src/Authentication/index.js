import { useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import getConfig from 'next/config'
import { SWRConfig } from 'swr'

import Login from '../Login'
import Projects from '../Projects'

import { getUser, authenticateUser, logout, fetcher } from './helpers'

const {
  publicRuntimeConfig: { GOOGLE_OAUTH_CLIENT_ID },
} = getConfig()

export const noop = () => () => {}

let googleAuth = { signIn: noop, signOut: noop }

const Authentication = ({ children }) => {
  const [hasLocalStorageLoaded, setHasLocalStorageLoaded] = useState(false)
  const [hasGoogleLoaded, setHasGoogleLoaded] = useState(false)
  const [errorMessage, setErrorMessage] = useState('')
  const [user, setUser] = useState({})

  useEffect(() => {
    window.onload = () => {
      window.gapi.load('auth2', async () => {
        googleAuth = window.gapi.auth2.init({
          client_id: `${GOOGLE_OAUTH_CLIENT_ID}`,
        })
        setHasGoogleLoaded(true)
      })
    }
  }, [])

  useEffect(() => {
    if (hasLocalStorageLoaded) return

    const storedUser = getUser()

    setUser(storedUser)

    setHasLocalStorageLoaded(true)
  }, [hasLocalStorageLoaded, user])

  if (!hasLocalStorageLoaded) return null

  if (!user.id) {
    return (
      <Login
        googleAuth={googleAuth}
        hasGoogleLoaded={hasGoogleLoaded}
        errorMessage={errorMessage}
        setErrorMessage={setErrorMessage}
        onSubmit={authenticateUser({ setErrorMessage, setUser })}
      />
    )
  }

  return (
    <SWRConfig value={{ fetcher: fetcher({ setUser }) }}>
      <Projects user={user} logout={logout({ googleAuth, setUser })}>
        {({ selectedProject }) =>
          children({
            user,
            logout: logout({ googleAuth, setUser }),
            selectedProject,
          })
        }
      </Projects>
    </SWRConfig>
  )
}

Authentication.propTypes = {
  children: PropTypes.func.isRequired,
}

export default Authentication
