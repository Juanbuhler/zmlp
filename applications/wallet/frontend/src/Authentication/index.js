import { useContext, useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import getConfig from 'next/config'
import { SWRConfig } from 'swr'
import Router from 'next/router'

import { initializeFetcher } from '../Fetch/helpers'

import { UserContext } from '../User'

import Login from '../Login'
import Projects from '../Projects'
import Layout from '../Layout'
import ErrorBoundary from '../ErrorBoundary'

import { authenticateUser, logout } from './helpers'

const AUTHENTICATION_LESS_ROUTES = ['/create-account', '/reset-password']

const {
  publicRuntimeConfig: { GOOGLE_OAUTH_CLIENT_ID },
} = getConfig()

export const noop = () => () => {}

const Authentication = ({ route, children }) => {
  const { user, setUser, googleAuth, setGoogleAuth } = useContext(UserContext)

  const [hasGoogleLoaded, setHasGoogleLoaded] = useState(false)
  const [errorMessage, setErrorMessage] = useState('')

  const fetcher = initializeFetcher({ setUser })

  useEffect(() => {
    window.onload = () => {
      window.gapi.load('auth2', async () => {
        setGoogleAuth(
          window.gapi.auth2.init({
            client_id: `${GOOGLE_OAUTH_CLIENT_ID}`,
          }),
        )
        setHasGoogleLoaded(true)
      })
    }
  }, [setGoogleAuth])

  if (AUTHENTICATION_LESS_ROUTES.includes(route)) {
    if (user.id) {
      Router.push('/')
      return null
    }

    return children
  }

  if (!user.id) {
    return (
      <Login
        googleAuth={googleAuth}
        hasGoogleLoaded={hasGoogleLoaded}
        errorMessage={errorMessage}
        setErrorMessage={setErrorMessage}
        onSubmit={authenticateUser({ setUser, setErrorMessage })}
      />
    )
  }

  return (
    <SWRConfig value={{ fetcher }}>
      <Projects projectId={user.projectId} setUser={setUser}>
        <Layout user={user} logout={logout({ googleAuth, setUser })}>
          <ErrorBoundary>{children}</ErrorBoundary>
        </Layout>
      </Projects>
    </SWRConfig>
  )
}

Authentication.propTypes = {
  children: PropTypes.node.isRequired,
  route: PropTypes.string.isRequired,
}

export default Authentication
