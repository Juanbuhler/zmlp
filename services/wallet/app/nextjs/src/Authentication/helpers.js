import jwtDecode from 'jwt-decode'

import { axiosCreate } from '../Axios/helpers'

import { ACCESS_TOKEN, REFRESH_TOKEN } from './constants'

export const getTokens = () => {
  if (typeof window === 'undefined') return {}

  const accessToken = localStorage.getItem(ACCESS_TOKEN)
  const refreshToken = localStorage.getItem(REFRESH_TOKEN)

  if (accessToken && refreshToken) {
    return { accessToken, refreshToken }
  }

  return {}
}

export const isUserAuthenticated = ({ refreshToken }) => {
  if (!refreshToken) return false

  const decodedToken = jwtDecode(refreshToken)
  const currentTime = Date.now() / 1000
  const expirationTime = decodedToken.exp

  const isAuthenticated = currentTime < expirationTime

  return isAuthenticated
}

export const clearTokens = () => {
  localStorage.removeItem(ACCESS_TOKEN)
  localStorage.removeItem(REFRESH_TOKEN)
}

export const storeTokens = ({ accessToken, refreshToken }) => {
  localStorage.setItem(ACCESS_TOKEN, accessToken)
  localStorage.setItem(REFRESH_TOKEN, refreshToken)
}

export const getTokenTimeout = ({ refreshToken }) => {
  const decodedToken = jwtDecode(refreshToken)
  const currentTime = Date.now() / 1000
  const expirationTime = decodedToken.exp

  return (expirationTime - currentTime - 30) * 1000 // set to 30 seconds before expiration
}

export const authenticateUser = ({ setUser }) => ({ username, password }) => {
  const instance = axiosCreate({})

  return instance
    .post('/auth/token/', { username, password })
    .then(({ data }) => {
      const { access: accessToken, refresh: refreshToken } = data
      storeTokens({ accessToken, refreshToken })
      setUser({ isAuthenticated: true })
    })
    .catch(error => console.error(error))
}

export const logout = ({ setUser }) => () => {
  clearTokens()
  setUser({ isAuthenticated: false })
}
