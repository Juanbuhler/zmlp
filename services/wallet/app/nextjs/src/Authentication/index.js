import { useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import { SWRConfig } from 'swr'

import { axiosCreate, fetcher } from '../Axios/helpers'

import Login from '../Login'

import {
  getTokens,
  isUserAuthenticated,
  authenticateUser,
  getTokenTimeout,
  logout,
} from './helpers'

const Authentication = ({ children }) => {
  const axiosInstance = axiosCreate({})

  const [user, setUser] = useState({ hasLoaded: false, isAuthenticated: false })

  useEffect(() => {
    let timeoutId

    const { refreshToken } = getTokens()

    if (!user.hasLoaded) {
      setUser({
        hasLoaded: true,
        isAuthenticated: isUserAuthenticated({ now: Date.now(), refreshToken }),
      })
    }

    if (user.isAuthenticated) {
      timeoutId = setTimeout(
        logout({ setUser }),
        getTokenTimeout({ now: Date.now(), refreshToken }),
      )
    }

    return () => clearTimeout(timeoutId)
  }, [user])

  if (!user.hasLoaded) return null

  if (!user.isAuthenticated) {
    return <Login onSubmit={authenticateUser({ axiosInstance, setUser })} />
  }

  return (
    <SWRConfig value={{ fetcher: fetcher({ axiosInstance }) }}>
      {children({ user, logout: logout({ setUser }) })}
    </SWRConfig>
  )
}

Authentication.propTypes = {
  children: PropTypes.func.isRequired,
}

export default Authentication
