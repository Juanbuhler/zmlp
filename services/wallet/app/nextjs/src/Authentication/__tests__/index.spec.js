import TestRenderer, { act } from 'react-test-renderer'

import Authentication from '../'

jest.mock('../../Login', () => 'Login')

describe('<Authentication />', () => {
  it('should render properly', () => {
    const component = TestRenderer.create(
      <Authentication>{({ user }) => `Hello ${user.email}!`}</Authentication>,
    )

    expect(component.toJSON()).toMatchSnapshot()

    act(() => {
      component.root.findByType('Login').props.onSubmit({ email: 'World' })()
    })

    expect(component.toJSON()).toMatchSnapshot()
  })
})
