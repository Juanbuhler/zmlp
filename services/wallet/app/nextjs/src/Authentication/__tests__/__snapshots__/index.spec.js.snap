// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Authentication /> should render properly 1`] = `
<Login
  onSubmit={[Function]}
/>
`;

exports[`<Authentication /> should render properly 2`] = `"Hello World!"`;
